util/tlgen/lib.py

```diff
@@ -28,37 +32,37 @@ def simplify_addr(dev: Dict[Any, Any],
     {0x4000_0000, 0x1_0000}, {0x4008_0000, 0x1_0000} then it combines two
     entries into {0x4000_0000, 0x9_0000}
 
-    @param addrs List of Dict[Addr] : {'base_addr':,'size_byte':}
+    @param addrs List of Dict[Addr] : {'base_addr':Dict[ASID,Addr],'size_byte':}
     """
 
     addrs = dev["addr_range"]
     # Sort based on the base addr
-    newlist = sorted(addrs, key=lambda k: int(k['base_addr'], 0))
+    newlist = sorted(addrs, key=lambda k: int(k['base_addrs'][asid], 0))
     # check if overlap or contiguous
     result: List[Dict[Any, Any]] = []
     for e in newlist:
         if len(result) == 0:
             result.append(e)
             continue
         # if contiguous
-        if int(e["base_addr"], 0) == int(result[-1]["base_addr"], 0) + int(
+        if int(e["base_addrs"][asid], 0) == int(result[-1]["base_addrs"][asid], 0) + int(
                 result[-1]["size_byte"], 0):
             # update previous entry size
             result[-1]["size_byte"] = "0x{:x}".format(
                 int(result[-1]["size_byte"], 0) + int(e["size_byte"], 0))
             continue
 
-        if no_device_in_range(xbar, dev["name"], result[-1], e):
+        if no_device_in_range(xbar, dev["name"], result[-1], e, asid):
             # Determine if size can be power of 2 value
-            next_value = (get_next_base_addr(e["base_addr"], xbar,
-                                             dev["name"]) or
+            next_value = (get_next_base_addr(e["base_addrs"], xbar,
+                                             dev["name"], asid) or
                           0x100000000)
 
-            calc_size = int(e["base_addr"], 0) + int(e["size_byte"], 0) - int(
-                result[-1]["base_addr"], 0)
+            calc_size = int(e["base_addrs"][asid], 0) + int(e["size_byte"], 0) - int(
+                result[-1]["base_addrs"][asid], 0)
 
             # find power of 2 if possible
-            size_byte = find_pow2_size(result[-1], calc_size, next_value)
+            size_byte = find_pow2_size(result[-1]["base_addrs"][asid], calc_size, next_value)
```

Comment:
The first arg of find_pow2_size seems to be a string?

Reply:
See the comment above.